ext/opcache/jit/zend_jit_x86.dasc

```diff
@@ -4610,14 +4621,29 @@ static int zend_jit_assign_dim(dasm_State **Dst, const zend_op *opline, const ze
 
 	if (op1_info & MAY_BE_REF) {
 		|	LOAD_ZVAL_ADDR FCARG1a, op1_addr
-		|	ZVAL_DEREF FCARG1a, op1_info
+		|	IF_NOT_Z_TYPE FCARG1a, IS_REFERENCE, >1
+		|	GET_Z_PTR FCARG2a, FCARG1a
+		|	IF_NOT_TYPE byte [FCARG2a + offsetof(zend_reference, val) + offsetof(zval, u1.v.type)], IS_ARRAY, >2
+		|	lea FCARG1a, [FCARG2a + offsetof(zend_reference, val)]
+		|	jmp >3
+		|.cold_code
+		|2:
+		|	SAVE_VALID_OPLINE opline
+		|	EXT_CALL zend_jit_prepare_assign_dim_ref, r0
+		|	test r0, r0
+		|	jz ->exception_handler_undef
+		|	mov FCARG1a, r0
+		|	jmp >1
+		|.code
+		|1:
```

Comment:
@dstogov How does this version look? It checks for array references and otherwise falls back to helper.